### PR TITLE
Implement Cross Resource Referencing

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplaneio/crossplane/issues/890
Replaces https://github.com/crossplaneio/stack-azure/pull/39

TODO:

* Clean up commit history
* Test it

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplaneio/stack-azure/blob/master/config/stack/manifests/app.yaml

### CODE REVIEW

File: apis/compute/v1alpha2/types.go [hunk 1]
```diff
@@ -34,12 +40,53 @@ const (
 	DefaultNodeCount = 1
 )
 
+// Error strings
+const (
+	errResourceIsNotAKSCluster = "The managed resource is not a AKSCluster"
```
Comment (hasheddan) [inline apis/compute/v1alpha2/types.go]:
```suggestion
	errResourceIsNotAKSCluster = "The managed resource is not an AKSCluster"
```

File: apis/compute/v1alpha2/types.go [hunk 2]
```diff
@@ -34,12 +40,53 @@ const (
...
+	return nil
+}
+
 // AKSClusterParameters define the desired state of an Azure Kubernetes Engine
 // cluster.
 type AKSClusterParameters struct {
 	// ResourceGroupName is the name of the resource group that the cluster will
 	// be created in
-	ResourceGroupName string `json:"resourceGroupName"`
+	ResourceGroupName string `json:"resourceGroupName,omitempty"`
```
Comment (hasheddan) [inline apis/compute/v1alpha2/types.go]: Should this have `// +optional` as well? It looks like it is on all `omitempty` fields except for the reference fields. I don't know exactly how the mechanics of the `// +optional` tag works, so I may be off-base here.

File: apis/database/v1alpha2/sql_types.go [hunk 3]
```diff
@@ -260,6 +311,26 @@ type VirtualNetworkRuleStatus struct {
...
+	Name string `json:"name"`
+
+	// ServerName - Name of the Virtual Network Rule's PostgresqlServer.
+	ServerName string `json:"serverName,omitempty"`
+
+	// ServerNameRef - A reference to the Virtual Network Rule's PostgresqlServer.
+	ServerNameRef *PostgresqlServerNameReferencer `json:"serverNameRef,omitempty" resource:"attributereferencer"`
+
+	// ResourceGroupName - Name of the Virtual Network Rule's resource group.
+	ResourceGroupName string `json:"resourceGroupName"`
```
Comment (hasheddan) [inline apis/database/v1alpha2/sql_types.go]: Do we want a `ResourceGroupNameRef` in addition to `ResourceGroupName`?

File: apis/database/v1alpha2/sql_types.go [hunk 4]
```diff
@@ -224,23 +292,6 @@ type VirtualNetworkRuleProperties struct {
 	IgnoreMissingVnetServiceEndpoint bool `json:"ignoreMissingVnetServiceEndpoint,omitempty"`
```
Comment (hasheddan) [inline apis/database/v1alpha2/sql_types.go]: Do we want a `VirtualNetworkSubnetIdRef` field in `VirtualNetworkRuleProperties`?

File: apis/network/v1alpha2/subnet_referencers.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+// Build retrieves and builds the NetworkName
+func (v *SubnetIDReferencer) Build(ctx context.Context, _ resource.CanReference, reader client.Reader) (string, error) {
+	refObj := Subnet{}
+	nn := types.NamespacedName{Name: v.Name}
+	if err := reader.Get(ctx, nn, &refObj); err != nil {
+		return "", err
+	}
+
+	return refObj.Status.ID, nil
```
Comment (hasheddan) [inline apis/network/v1alpha2/subnet_referencers.go]: Does this field return a value like `/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/demo-rg/providers/Microsoft.Network/virtualNetworks/demo-vnet/subnets/demo-subnet`?

File: apis/v1alpha2/types.go [hunk 6]
```diff
@@ -71,7 +71,7 @@ type ResourceGroupStatus struct {
 	// TODO(negz): Do we really need the name in both spec and status?
```
Comment (hasheddan) [inline apis/v1alpha2/types.go]: Looks like you are solving this by adding `omitempty`, okay to remove `TODO`?

File: apis/compute/v1alpha2/types.go (hunk 1 above)
Comment (negz) [reply apis/compute/v1alpha2/types.go]: While we're at it, error strings should not start with capitals in Go.

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: apis/database/v1alpha2/sql_types.go (hunk 3 above)
Comment (negz) [reply apis/database/v1alpha2/sql_types.go]: Good catch - I imagine we do.

File: apis/v1alpha2/types.go (hunk 6 above)
Comment (negz) [reply apis/v1alpha2/types.go]: I think the omitempty is something @soorena776 added due to issues serialising statuses? I'd prefer to leave the TODO as I think it's still relevant.

File: apis/compute/v1alpha2/types.go (hunk 2 above)
Comment (negz) [reply apis/compute/v1alpha2/types.go]: I avoided adding `+optional` comment markers to types that don't already use them - we should probably add them wholesale during a v1beta1 port.

File: apis/v1alpha2/types.go (hunk 6 above)
Comment (hasheddan) [reply apis/v1alpha2/types.go]: Sounds reasonable to me!

File: apis/compute/v1alpha2/types.go (hunk 1 above)
Comment (soorena776) [reply apis/compute/v1alpha2/types.go]: fixed these in the referencing #39

File: apis/database/v1alpha2/sql_types.go (hunk 3 above)
Comment (soorena776) [reply apis/database/v1alpha2/sql_types.go]: fixed in #39

File: apis/database/v1alpha2/sql_types.go (hunk 4 above)
Comment (soorena776) [reply apis/database/v1alpha2/sql_types.go]: We do. fixed in #39

File: apis/network/v1alpha2/subnet_referencers.go (hunk 5 above)
Comment (soorena776) [reply apis/network/v1alpha2/subnet_referencers.go]: it does :)

**Comment by negz** [comment]
Closing in favor of https://github.com/crossplaneio/stack-azure/pull/39.

**Pull request closed without merging**